chain/ethereum/src/block_ingestor.rs

```diff
@@ -73,40 +73,69 @@ where
         })
     }
 
-    pub fn into_polling_stream(self) -> impl Future<Item = (), Error = ()> {
+    pub fn into_polling_stream(self, cleanup_freq: Duration) -> impl Future<Item = (), Error = ()> {
         // Currently, there is no way to stop block ingestion, so just leak self
         let static_self: &'static _ = Box::leak(Box::new(self));
 
+        let mut last_cleanup = Instant::now();
+
         // Create stream that emits at polling interval
         tokio::timer::Interval::new(Instant::now(), static_self.polling_interval)
             .map_err(move |e| {
                 error!(static_self.logger, "timer::Interval failed: {:?}", e);
             })
             .for_each(move |_| {
                 // Attempt to poll
-                static_self.do_poll().then(move |result| {
-                    if let Err(err) = result {
-                        // Some polls will fail due to transient issues
-                        match err {
-                            EthereumAdapterError::BlockUnavailable(_) => {
-                                trace!(
-                                    static_self.logger,
-                                    "Trying again after block polling failed: {}",
-                                    err
-                                );
+                static_self
+                    .do_poll()
+                    .then(move |result| {
+                        if let Err(err) = result {
+                            // Some polls will fail due to transient issues
+                            match err {
+                                EthereumAdapterError::BlockUnavailable(_) => {
+                                    trace!(
+                                        static_self.logger,
+                                        "Trying again after block polling failed: {}",
+                                        err
+                                    );
+                                }
+                                EthereumAdapterError::Unknown(inner_err) => {
+                                    warn!(
+                                        static_self.logger,
+                                        "Trying again after block polling failed: {}", inner_err
+                                    );
+                                }
                             }
-                            EthereumAdapterError::Unknown(inner_err) => {
-                                warn!(
+                        }
+
+                        // Continue polling even if polling failed
+                        future::ok(())
+                    })
+                    .then(move |result| {
```

Comment:
You could use `inspect` here, then you don't need to pass `result` through:
```rust
.inspect(move |_| {
  ...
})
```
I do wonder though, wouldn't this better be implemented via a separate Tokio task / thread?

Reply:
What's the concern here? When you do cleanups, the overall slowdown of the system because you need to continually download blocks will far outweigh any slowdown from adding a `delete` query into the blockstream.

Reply:
I'm not sure I understand the "continually download blocks" part. I was just wondering if running this delete query every time there is a new chain head would slow down the ingestion, but since we can configure how frequently that happens, I think we're good here. It was more a question around a separation of concerns.